src/auto-ssh-agent/pam_auto_ssh_agent.c

```diff
@@ -0,0 +1,756 @@
+              free_safe (ret);
+              errno = ENOMEM;
+              return NULL;
+            }
+
+          memset (n + len, 0, r + 1);
+          ret = n;
+          len = len + r;
+          strncat (ret, buf, r);
+        }
+
+      if (r == 0 || len > MAX_LENGTH || consume == 0)
+        break;
+    }
+
+  return ret;
+}
+
+static int
+write_string (int fd,
+              const char* buf)
+{
+  size_t bytes = 0;
+  int res, len = strlen (buf);
+
+  while (bytes < len)
+    {
+      res = write (fd, buf + bytes, len - bytes);
+      if (res < 0)
+        {
+          if (errno != EINTR && errno != EAGAIN)
+            return -1;
+        }
+      else
+        {
+          bytes += res;
+        }
+    }
+
+  return 0;
+}
+
+static int
+log_problem (char *line,
+             void *arg)
+{
+  /*
+   * Called for each stderr output line from the daemon.
+   * Send it all to the log.
+   */
+
+  int *success;
+
+  assert (line);
+  assert (arg);
+
+  success = (int*)arg;
+  if (*success)
+    message ("%s", line);
+  else
+    error ("%s", line);
+
+  return PAM_SUCCESS;
+}
+
+static int
+parse_args (int argc,
+            const char **argv)
+{
+  int args = 0;
+  int i;
+
+  verbose_mode = 0;
+
+  /* Parse the arguments */
+  for (i = 0; i < argc; i++)
+    {
+      if (strcmp (argv[i], "verbose") == 0)
+        {
+          verbose_mode = 1;
+        }
+      else
+        {
+          message ("invalid option: %s", argv[i]);
+          continue;
+        }
+    }
+
+  return args;
+}
+
+static int
+setup_pam_env (pam_handle_t *pamh,
+               const char *name,
+               const char *val)
+{
+  int ret;
+  char *var;
+
+  assert (name);
+  assert (val);
+
+  var = malloc (strlen (name) + strlen (val) + 2);
+  if (!var)
+    {
+      error ("out of memory");
+      return PAM_SYSTEM_ERR;
+    }
+
+  sprintf (var, "%s=%s", name, val);
+  ret = pam_putenv (pamh, var);
+  free_safe (var);
+
+  return ret;
+}
+
+static void
+setup_child (int inp[2],
+             int outp[2],
+             int errp[2],
+             pam_handle_t *pamh,
+             struct passwd *pwd,
+             char **args)
+{
+  const char* display;
+  const char *runtime;
+
+  int i, ret;
+
+  assert (pwd);
+  assert (pwd->pw_dir);
+
+  /* Fix up our end of the pipes */
+  if (dup2 (inp[READ_END], STDIN) < 0 ||
+      dup2 (outp[WRITE_END], STDOUT) < 0 ||
+      dup2 (errp[WRITE_END], STDERR) < 0)
+    {
+      error ("couldn't setup pipes: %s", strerror (errno));
+      exit (EXIT_FAILURE);
+    }
+
+  /* Try valiantly to close unnecessary file descriptors */
+  for (i = STDERR + 1; i < 64; ++i)
+    close (i);
+
+  /* Close unnecessary file descriptors */
+  close (inp[READ_END]);
+  close (inp[WRITE_END]);
+  close (outp[READ_END]);
+  close (outp[WRITE_END]);
+  close (errp[READ_END]);
+  close (errp[WRITE_END]);
+
+  /* We may be running effective as another user, revert that */
+  if (seteuid (getuid ()) < 0 || setegid (getgid ()) < 0)
```

Comment:
This needs to happen in the opposite order, set group first, and then drop user permissions.